### PR TITLE
Add tests for `_fast_non_dominated_sort`

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
Currently, the tests for `_fast_non_dominated_sort` misses some important edge cases. The missed edge cases include
* `inf` and `-inf` in values/constraints
* `nan` in constraints
* duplicated values/constraints
* apparently different constraint values with same meaning (e.g. both are feasible)
etc.
We refactor the tests and cover these cases.

Random tests, suggested in #3678, is not included in this PR.

This is a follow up PR after #3683.

## Description of the changes
* Add test cases.
* Add a function to automatically judge whether the output is correct, i.e., all trials in the same rank do not dominate each other and all trials in rank `>0` are dominated by some trial in the rank above.

### CODE REVIEW

**Comment by contramundum53** [comment]
@HideakiImamura @c-bata Could you review this PR?

File: tests/samplers_tests/test_nsgaii.py [hunk 1]
```diff
@@ -304,266 +306,115 @@ def test_constrained_dominates_infeasible_vs_infeasible(direction: StudyDirectio
...
-        {0, 1},
-        {3, 4, 5},
-        {2},
-    ]
+    values = itertools.product(
+        [10, 20, 20, 30, float("inf"), float("inf"), -float("inf")], repeat=values_dim
+    )
+    for directions in itertools.product(
+        [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE], repeat=values_dim
+    ):
```
Comment (HideakiImamura) [inline tests/samplers_tests/test_nsgaii.py:345]: Can we use `pytest.mark.parametrize` for this?

File: tests/samplers_tests/test_nsgaii.py [hunk 2]
```diff
@@ -304,266 +306,115 @@ def test_constrained_dominates_infeasible_vs_infeasible(direction: StudyDirectio
...
-        _create_frozen_trial(4, [15, 15], [4, 4]),
-    ]
-    population_per_rank = sampler._fast_non_dominated_sort(trials, directions)
-    assert [{t.number for t in population} for population in population_per_rank] == [
-        {0},
-        {1},
-        {2, 3},
-        {4},
-    ]
+@pytest.mark.parametrize("values_dim", [1, 2, 3])
```
Comment (HideakiImamura) [inline tests/samplers_tests/test_nsgaii.py:336]: How about splitting each dimension test into separate functions as we done for `_constrained_dominates`?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the update. I have comments. PTAL.

File: tests/samplers_tests/test_nsgaii.py (hunk 2 above)
Comment (contramundum53) [reply tests/samplers_tests/test_nsgaii.py:336]: Unlike `_constrained_dominates`, this does not make the test easier, since we need to get the whole list anyways.

**Comment by contramundum53** [comment]
Sorry for late response. I foxed the test.

**Comment by contramundum53** [comment]
@HideakiImamura @c-bata PTAL.

**Comment by HideakiImamura** [comment]
Could you merge the master to fix the CI error?

**Comment by contramundum53** [comment]
@HideakiImamura Fixed. PTAL.

**Comment by contramundum53** [comment]
I simplified `test_fast_non_dominated_sort_with_constraints`.

File: tests/samplers_tests/test_nsgaii.py [hunk 3]
```diff
@@ -336,266 +338,111 @@ def test_constrained_dominates_infeasible_vs_infeasible(direction: StudyDirectio
...
-        _create_frozen_trial(5, [0, 9, 9], [0]),
-    ]
-    population_per_rank = sampler._fast_non_dominated_sort(trials, directions)
-    assert [{t.number for t in population} for population in population_per_rank] == [
-        {2},
-        {0, 3, 5},
-        {1},
-        {4},
-    ]
+                    assert not _constrained_dominates(trial1, trial2, direction)
```
Comment (c-bata) [inline tests/samplers_tests/test_nsgaii.py:357]: I'm wondering if these lines can check any behavior of `_non_dominated_sort` since `trial1` and `trial2` are completely the same `FrozenTrial` objects. What do you think?

File: tests/samplers_tests/test_nsgaii.py [hunk 4]
```diff
@@ -336,266 +338,111 @@ def test_constrained_dominates_infeasible_vs_infeasible(direction: StudyDirectio
...
 
+        # Check that each trial is dominated by some trial in the rank above.
+        for i in range(len(population_per_rank) - 1):
+            for trial2 in population_per_rank[i + 1]:
+                assert any(
+                    _constrained_dominates(trial1, trial2, direction)
+                    for trial1 in population_per_rank[i]
+                )
```
Comment (c-bata) [inline tests/samplers_tests/test_nsgaii.py:365]: Do you think we can write these logic like the following since we've already checked that trials in the same rank do not dominate each other?

```suggestion
        for i in range(len(population_per_rank) - 1):
            trial1 = population_per_rank[i][0]
            for trial2 in population_per_rank[i + 1]:
                assert _constrained_dominates(trial1, trial2, direction)
```

**Comment by c-bata** [comment]
Thank you for your pull request. Changes basically looks good except for my previous comments.

File: tests/samplers_tests/test_nsgaii.py (hunk 4 above)
Comment (contramundum53) [reply tests/samplers_tests/test_nsgaii.py:365]: No, you cannot write it this way because a trial in a lower rank may not dominate a particular trial in higher rank. 
An example is
* Rank 0: `(0, 5)`, `(5, 0)`
* Rank 1: `(7, 1)`
Here, `(7, 1)` is of rank 1 because `(7, 1)` is dominated by `(5, 0)`, but it is not dominated by `(0, 5)`.

Comment (c-bata) [reply tests/samplers_tests/test_nsgaii.py:365]: I see. Thank you for your explanation :bow:

**Review by c-bata — APPROVED** [review]
LGTM after passed CI.

**Review by HideakiImamura — APPROVED** [review]
LGTM.
